k8s/k8s.go

```diff
@@ -250,26 +264,141 @@ func CmdAddK8s(args *skel.CmdArgs, conf utils.NetConf, hostname string, calicoCl
+				return nil, fmt.Errorf("Error getting IP from IPAM: %s", err)
+			}
+
+			// Set the IP6 part of the result from the type.Result returned by the IPAM plugin.
+			result.IP6 = r.IP6
+			logger.Debugf("Adding IPv6: %s to result", ipAddr.String())
+		} else {
+			// It's an IPv4 address.
+			if result.IP4 != nil {
+				logger.Error("Can not have more than one IPv4 address in ipAddrs annotation")
+				return nil, fmt.Errorf("Can not have more than one IPv4 address in ipAddrs annotation")
+			}
+
+			// Call callIPAMWithIP with the ip address.
+			r, err := callIPAMWithIP(ipAddr, conf, args, logger)
+			if err != nil {
+				return nil, fmt.Errorf("Error getting IP from IPAM: %s", err)
+			}
+
+			// Set the IP4 part of the result from the type.Result returned by the IPAM plugin.
+			result.IP4 = r.IP4
+			logger.Debugf("Adding IPv4: %s to result", ipAddr.String())
+		}
+	}
+
+	return &result, nil
+}
+
+// callIPAMWithIP sets CNI_ARGS with the IP and calls the IPAM plugin with it
+// to get types.Result and then it unsets the IP field from CNI_ARGS ENV var,
+// so it doesn't pollute the subsequent requests.
+func callIPAMWithIP(ip net.IP, conf utils.NetConf, args *skel.CmdArgs, logger *log.Entry) (res *types.Result, finalErr error) {
+
+	// Save the original value of the CNI_ARGS ENV var for backup.
+	originalArgs := os.Getenv("CNI_ARGS")
+	logger.Debugf("Original CNI_ARGS=%s", originalArgs)
+
+	ipamArgs := struct {
+		types.CommonArgs
+		IP net.IP `json:"ip,omitempty"`
+	}{}
+
+	if err := types.LoadArgs(args.Args, &ipamArgs); err != nil {
+		return nil, err
+	}
+
+	if ipamArgs.IP != nil {
+		logger.Errorf("'IP' variable already set in CNI_ARGS environment variable.")
+	}
+
+	// Request the provided IP address using the IP CNI_ARG.
+	// See: https://github.com/containernetworking/cni/blob/master/CONVENTIONS.md#cni_args for more info.
+	newArgs := originalArgs + ";IP=" + ip.String()
+	logger.Debugf("New CNI_ARGS=%s", newArgs)
+
+	// Set CNI_ARGS to the new value.
+	err := os.Setenv("CNI_ARGS", newArgs)
+	if err != nil {
+		return nil, fmt.Errorf("Error setting CNI_ARGS environment variable: %v", err)
+	}
+
+	// Restore the CNI_ARGS ENV var to it's original value,
+	// so the subsequent calls don't get polluted by the old IP value.
+	defer func() {
+		err := os.Setenv("CNI_ARGS", originalArgs)
+		if err != nil {
+			// Need to clean up IP allocation if this step doesn't succeed.
+			utils.ReleaseIPAllocation(logger, conf.IPAM.Type, args.StdinData)
+
+			// Returned values from a defered closure func get ignored in Go, so have to use named return.
+			// Following two lines are equivalent of return nil, fmt.Errorf("Error setting CNI_ARGS environment variable: %v", err)
+			res = nil
+			finalErr = fmt.Errorf("Error setting CNI_ARGS environment variable: %v", err)
+		}
+	}()
+
+	// Run the IPAM plugin.
+	logger.Debugf("Calling IPAM plugin %s", conf.IPAM.Type)
+	r, err := ipam.ExecAdd(conf.IPAM.Type, args.StdinData)
+	if err != nil {
+		return nil, err
```

Comment:
I had to do named return because defere'd closure func in Go gets ignored, so only way to pass it to the parent function's return is to use named returns. See: https://groups.google.com/forum/#!msg/golang-nuts/qTTBENO_Em0/Y4MCVZZ3c5sJ (second message)